geode-core/src/main/java/org/apache/geode/internal/cache/wan/parallel/ParallelGatewaySenderQueue.java

```diff
@@ -1404,13 +1404,19 @@ void peekEventsFromIncompleteTransactions(List<GatewaySenderEventImpl> batch,
         }
       }
       if (incompleteTransactionIdsInBatch.size() == 0 ||
-          retries++ == GET_TRANSACTION_EVENTS_FROM_QUEUE_RETRIES) {
+          retries >= sender.getRetriesToGetTransactionEventsFromQueue()) {
         break;
       }
+      retries++;
```

Comment:
I am assuming this is a best effort to find missing transaction events that are getting added to the Queue. If thats the case, can it be done based on wait/notify on Queue? Right now the default wait is for 1ms; but in case if someone sets it longer than wait/notify will be efficient.
Also why is the default 1ms; isn't it too small...

Reply:
Yes, as you say, this is a best effort.
Given that setting a high value for the wait could lead to delaying too much the replication, which could eventually provoke that queues grow indefinitely, I opted for setting a low value.
Also, in case of a misconfiguration (one not detected at configuration time) due to having transactions with events going to different gateway senders, it could be that it is impossible for the batch to contain the complete transaction. In that case, the maximum time will be consumed (retries*wait_time).

Regarding using wait/notify, I have not thought about it that much. I guess we could call notify when an element is added to the queue and call wait instead of sleep. But, if it makes sense to have a small sleep, then using wait and notify may be inefficient. Do you think it is worth to change it to wait/notify?

Reply:
Wait and notify is preferred over "sleep"; as sleep is not accurate/efficient way of waiting for an event. You can see if its possible to do...

Reply:
@agingade I have given a bit of thought about using wait and notify instead of sleep but I have come to the conclusion that it is not a good idea in this case.
According to your suggestion, instead of `sleep(time)` we could call `wait(timeout)` and then `notify()` whenever an event is added to the queue. This way, every event added to the queue will awake the thread that must look for events in the queue to complete transactions inside a batch.
The problem is that the event added to the queue that will provoke the awakening of the waiting thread may not be the one required to complete the transaction so the thread could have been awaken in vain.
If a lot of events are entering in the queue at a high rate, it might be that, a lot of awakes will be required before the events to complete the transaction enter the queue. This would require to increase the value of the retries. But this is dangerous as the WAN replication rate could greatly decrease.

The original implementation of this feature did not contemplate any sleeps. It just retried several times the search in the queue to retrieve events to complete transactions when they were not found, assuming that the events for a transaction should arrive to the queue very close in time. I observed that sometimes the retries were done so quickly that even if the number of retries was high, still some batches had to be sent with incomplete transactions and that is why I changed the behavior in this PR to wait a little before retrying by means of sleep.

If you do not agree with my conclusions, I am open to discuss them with you by e-mail, chat or videoconference.

Reply:
I feel it's good to sleep 1ms, but retry number might be better to keep 10 instead of 3.

Reply:
GET_TRANSACTION_EVENTS_FROM_QUEUE_WAIT_TIME_MS is 1ms by default.

GET_TRANSACTION_EVENTS_FROM_QUEUE_RETRIES is 3 by default.

So, thats 3 ms of retries by default. As Gester said, I'm not sure thats going to be enough.